src/DocumentDB.ChangeFeedProcessor/LeaseManagement/DocumentServiceLeaseManager.cs

```diff
@@ -27,36 +27,31 @@ namespace Microsoft.Azure.Documents.ChangeFeedProcessor.PartitionManagement
     internal class DocumentServiceLeaseManager : ILeaseManager
     {
         private static readonly ILog Logger = LogProvider.GetCurrentClassLogger();
-        private readonly string containerNamePrefix;
-        private readonly DocumentCollectionInfo leaseStoreCollectionInfo;
-        private readonly IChangeFeedDocumentClient client;
-        private readonly IDocumentServiceLeaseUpdater leaseUpdater;
-        private readonly string leaseStoreCollectionLink;
-        private readonly string hostName;
-
-        public DocumentServiceLeaseManager(
-            IChangeFeedDocumentClient client,
-            IDocumentServiceLeaseUpdater leaseUpdater,
-            DocumentCollectionInfo leaseStoreCollectionInfo,
```

Comment:
I would propose dropping LeaseManagerParameters -> passing dependencies via such "parameters" holder class is strange. 

I'm looking at the feature from this perspective. We need to extend API in such way, that e.g. I'm able to replace SinglePartitionLeaseManager to RedisLeaseManager or MultipartitonLeaseManager(PK:owner) with just 1 method CFBuilder.WithLeaseManager(custom_lease_manager). Or I'm able to create compound LeaseManager so that migratign from single to multiparition lease manager is smooth.

In order to do this I propose the following:
1) make DocumentServiceLeaseManager public and pass there all required dependencies -> client, leases settings (link to collection, prefix, hostname, link), prefix. Then, in order to create OwnerPartitionedDocumentServiceLeaseManager I would need to derive from this base and implement 2 factory methods for creating requests options.
2) ILeaseManager should include several role interfaces: ILeaseStore, ILeasesLoader, ILeaseOwnershipManager so that all the lease manipulation is inside 1 class but exposed by different so called "role interfaces". So that the bootstrapper uses ILeaseStore, LoadBalancer only ILeasesLoader, Controller just ILeaseOwnershipManager.  By doing this, the code for creating initial leases and leases after split should be same, just exposed by different role interfaces.
3) LeaseUpdater code would be included inside DocServiceLeaseManager because LeaseUpdater handles retry logic only over IDocClient with manipulating passed lease.

Reply:
CFPBuiler.WithLeaseManager is already there and is public. My original intent was only to expose ILeaseManager as public, and keep the implementations internal (otherwise we would need to make sure impl is clean as changes would be breaking changes). Can you provide an example how you are going to reuse/derive from DocumentServiceLeaseManager?

Reply:
I used parameter object pattern to make future extensibility easier. Adding new constructors is not great.

Reply:
WRT IRequestOptions originally I even wanted to keep it inside to DocumentServiceLeaseManager so that it would be visible on only to derived classes. But that would be harder for testing, so I put it at the same level. Potentially we can make it public but again I am not sure if it would need something else in recent future.

Reply:
Let's discuss this online.

Reply:
Done!